### PR TITLE
[DPE-2561] POC - Start with provided config option: config-server, shard

### PR DESCRIPTION
**Pull request by MiaAltieri**
## Overview
This PR enables the user to pass the following config options: `replication`(default), `shard`, and `config-server`. This PR supports this options by starting the charm as a `shard` or a `config-server`. If the config option `config-server` is provided, then the charm starts an internal `mongos` service that runs on 0.0.0.0 and is configured to the provided config server. 

## Testing 
As a POC there are not tests included in this PR. Testing was performed by hand with:
```
# deploy
charmcraft pack
juju deploy ./*charm --config role="shard" shard-one
juju deploy ./*charm --config role="config-server" config-server

# watch deployment and verify all hooks pass without error and charm is in active and idle state
watch -n1 --color juju status --color

# verify running correctly
juju ssh shard-one/0
systemctl status snap.charmed-mongodb.mongod.service
exit
juju ssh config-server/0
systemctl status snap.charmed-mongodb.mongod.service
systemctl status snap.charmed-mongodb.mongos.service
exit
```
## Accommodating for PBM changes in new snap revision

New snap revison packages a new version of the PBM tool which updated how errors were handled when querying PBM status, commit [4bf9d5f9608976cf84d813f521826ea25eb9e04b](https://github.com/canonical/mongodb-operator/pull/246/commits/4bf9d5f9608976cf84d813f521826ea25eb9e04b) reflects these necessary changes 

## Follow Up PR
Follow up PR is to be made immediately after merging of this feature is completed and is a requirement of finishing [DPE-2561](https://warthogs.atlassian.net/browse/DPE-2561). For this PR this will be starting `mongos` with a `--auth` and the same `--keyFile` used to start the`mongod` service

## Future PR
Future PR is to be started after the follow up PR has been made. Once `mongos` is started with auth [DPE-2562](https://warthogs.atlassian.net/browse/DPE-2562) will be started. This includes creating a basic shared library between `config-server` and `shard` components. In this PR we will:
- implement keyfile sharing across shard and config server components 
- implement adding shards to cluster

(Shard removal is saved for later.)

## POC 
As a POC this PR doesn't include handling edge cases, intelligent status reporting, unit tests, or integration tests. These will be handled later on down the line. Specifically:
1. block config change events for changing role of charm 
2. report status of internal `mongos` in `update_status`
3. unit tests and integration tests

[DPE-2561]: https://warthogs.atlassian.net/browse/DPE-2561?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/charms/mongodb/v0/helpers.py [hunk 1]
```diff
@@ -80,10 +80,36 @@ def get_create_user_cmd(
     ]
 
 
+def get_mongos_args(config: MongoDBConfiguration) -> str:
```
Comment (MiaAltieri) [inline lib/charms/mongodb/v0/helpers.py:83]: bump lib patch plz

**Review by MiaAltieri — COMMENTED** [review]
WAITING TO MERGE - requires that `charmed-mongodb 5/edge` has been pushed to stable.

**Review by delgod — CHANGES_REQUESTED** [review]

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 2]
```diff
@@ -652,3 +647,48 @@ def _get_backup_restore_operation_result(self, current_pbm_status, previous_pbm_
...
+        replicas in the set.
+        """
+        try:
+            clusters = pbm_status["cluster"]
+            for cluster in clusters:
+                if cluster["rs"] == self.charm.app.name:
+                    break
+
+            for host_info in cluster["nodes"]:
+                replica_info = f"mongodb/{self.charm._unit_ip(self.charm.unit)}:27107"
```
Comment (dmitry-ratushnyy) [inline lib/charms/mongodb/v0/mongodb_backups.py:665]: It is possible to use some constant for the port value, or even better config.py?

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 3]
```diff
@@ -652,3 +647,48 @@ def _get_backup_restore_operation_result(self, current_pbm_status, previous_pbm_
...
+                if host_info["host"] == replica_info:
+                    break
+
+            return str(host_info["errors"])
+        except KeyError:
+            return ""
+
+    _StrOrBytes = Union[str, bytes]
```
Comment (dmitry-ratushnyy) [inline lib/charms/mongodb/v0/mongodb_backups.py:673]: nits: could we move type declaration to the top of the file?

File: src/charm.py [hunk 4]
```diff
@@ -950,12 +959,17 @@ def _initialise_replica_set(self, event: StartEvent) -> None:
...
+
                 logger.info("User initialization")
                 self._init_operator_user()
                 self._init_backup_user()
                 self._init_monitor_user()
-                logger.info("Manage relations")
-                self.client_relations.oversee_users(None, None)
+
+                # in sharding, user management is handled by mongos subordinate charm
+                if self.is_role(Config.REPLICATION):
```
Comment (dmitry-ratushnyy) [inline src/charm.py:969]: nits: maybe we can use a constant named "REPLICA" or similar (because "REPLICATION" looks like action) ?

File: src/config.py [hunk 5]
```diff
@@ -9,14 +9,16 @@
...
-    # We expect the MongoDB container to use the default ports
     MONGODB_PORT = 27017
+    SUBSTRATE = "vm"
     ENV_VAR_PATH = "/etc/environment"
     MONGODB_SNAP_DATA_DIR = "/var/snap/charmed-mongodb/current"
     MONGOD_CONF_DIR = f"{MONGODB_SNAP_DATA_DIR}/etc/mongod"
     MONGOD_CONF_FILE_PATH = f"{MONGOD_CONF_DIR}/mongod.conf"
-    SNAP_PACKAGES = [("charmed-mongodb", "5/edge", 82)]
+    SNAP_PACKAGES = [("charmed-mongodb", "5/edge", 84)]
+    CONFIG_SERVER = "config-server"
```
Comment (dmitry-ratushnyy) [inline src/config.py:19]: Maybe we can put CONFIG_SERVER, REPLICATION and SHARD into "ROLE" subclass or something like that?

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 2 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/mongodb_backups.py:665]: is this correct start of URI: "mongodb/"?, looks like something is missing

File: src/charm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (delgod) [reply src/charm.py:969]: Sorry, the current code is correct, because it is external user management.
However internal user management is still not protected and can be handled as separate PR.

File: src/config.py [hunk 6]
```diff
@@ -9,14 +9,16 @@
 class Config:
     """Configuration for MongoDB Charm."""
 
-    SUBSTRATE = "vm"
-    # We expect the MongoDB container to use the default ports
     MONGODB_PORT = 27017
+    SUBSTRATE = "vm"
     ENV_VAR_PATH = "/etc/environment"
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/config.py:14]:
```
cat /etc/systemd/system/snap.charmed-mongodb.mongos.service
[Unit]
# Auto-generated, DO NOT EDIT
Description=Service for snap application charmed-mongodb.mongos
Requires=snap-charmed\x2dmongodb-84.mount
Wants=network.target
After=snap-charmed\x2dmongodb-84.mount network.target snapd.apparmor.service
X-Snappy=yes

[Service]
EnvironmentFile=-/etc/environment
ExecStart=/usr/bin/snap run charmed-mongodb.mongos
SyslogIdentifier=charmed-mongodb.mongos
Restart=always
RestartSec=20
WorkingDirectory=/var/snap/charmed-mongodb/84
TimeoutStopSec=30
Type=simple

[Install]
WantedBy=multi-user.target
```

File: src/charm.py (hunk 4 above)
Comment (MiaAltieri) [reply src/charm.py:969]: I was keeping the name REPLICATION because it is consistent with the spec doc's naming of roles. (i.e. the role name is `replication`, so I think the config option should use the should use the same name

File: lib/charms/mongodb/v0/helpers.py [hunk 7]
```diff
@@ -80,10 +81,37 @@ def get_create_user_cmd(
...
+    """
+    # mongos running on the config server communicates through localhost
+    config_server_uri = f"{config.replset}/localhost"
+
+    # no need to add TLS since no network calls are used, since mongos is configured to listen
+    # on local host
+    cmd = [
+        # mongos on config server side should run on 0.0.0.0 so it can be accessed by other units
+        # in the sharded cluster
+        "--bind_ip 0.0.0.0",
```
Comment (delgod) [inline lib/charms/mongodb/v0/helpers.py:98]: lets enable IPv6 too
```suggestion
        "--bind_ip_all",
```

**Review by delgod — APPROVED** [review]

**Review by dmitry-ratushnyy — DISMISSED** [review]

**Review by dmitry-ratushnyy — DISMISSED** [review]

**Review by dmitry-ratushnyy — APPROVED** [review]

**Review by delgod — APPROVED** [review]
